### ISSUE TITLE
Avoiding the Despill Step

### ISSUE THREAD

**Issue by LinuxinaBit** [nikopueringer/CorridorKey#74]
It might be easy to almost entirely avoid the despill step by using different training data.

From the [video on Corridor Crew](https://youtu.be/3Ploi723hg4) it looks like the current training data is coming from a comp of a render onto a background plate rather than a combined render of both the subject and background.

It *would* take a lot of extra compute during training, but if the training data used a second full render of both the background and subject, *instead of compositing*, then the combined render would have spill from the background on the subject that CorridorKey would then be trained to reverse.

| Current training data | Training data that removes spill |
| ------------- | ------------- |
| Background plate | Background plate |
| Subject-only render | Subject-only render |
| Subject only render composited onto background plate |  Combined render of subject in front of background plate |

**Comment by SystemDisc** [comment]
This was my first thought when I saw the video. Rendering with the background and with reflections would give better training data.

**Comment by sarev** [comment]
Yes. I commented the same on YouTube, copying here for more likelihood of being spotted - there's also a follow-up idea related to training for shadows:

After you present the NN with the render to segment, which will have the foreground and the green-screen, and all of the green bounce lighting, you need to render the foreground without the ambient/reflected (green) light from the green-screen, i.e. in the ideal mid-grey environment, as your ground truth foreground image. You can already remove the foreground to just render the green-screen as the ground truth background image. With these two training sets, the NN should be able to more accurately predict what the subject should look like if it weren't surrounded by green surfaces. So your NN will be able to make better (good) guesses at undoing some of that green spill you always get on the foreground objects.

You can also do a similar trick in that the ground truth foreground element could be rendered with nothing behind it, so your shadows aren't hitting the green-screen, they are correctly cast onto the floor. This may help the NN to start to understand about casting shadows as a part of the prediction. But I'd be tempted to treat shadows as a separate system - if you can create an NN that, given a "no shadows" foreground render and a ground truth shadow mask render, it can start to predict shadow masks.
